pkg/query-service/app/traces/v3/query_builder.go

```diff
@@ -0,0 +1,385 @@
+	v3 "go.signoz.io/signoz/pkg/query-service/model/v3"
+	"go.signoz.io/signoz/pkg/query-service/utils"
+)
+
+var aggregateOperatorToPercentile = map[v3.AggregateOperator]float64{
+	v3.AggregateOperatorP05: 0.05,
+	v3.AggregateOperatorP10: 0.10,
+	v3.AggregateOperatorP20: 0.20,
+	v3.AggregateOperatorP25: 0.25,
+	v3.AggregateOperatorP50: 0.50,
+	v3.AggregateOperatorP75: 0.75,
+	v3.AggregateOperatorP90: 0.90,
+	v3.AggregateOperatorP95: 0.95,
+	v3.AggregateOperatorP99: 0.99,
+}
+
+var aggregateOperatorToSQLFunc = map[v3.AggregateOperator]string{
+	v3.AggregateOperatorAvg:     "avg",
+	v3.AggregateOperatorMax:     "max",
+	v3.AggregateOperatorMin:     "min",
+	v3.AggregateOperatorSum:     "sum",
+	v3.AggregateOperatorRate:    "count",
+	v3.AggregateOperatorRateSum: "sum",
+	v3.AggregateOperatorRateAvg: "avg",
+	v3.AggregateOperatorRateMax: "max",
+	v3.AggregateOperatorRateMin: "min",
+}
+
+var tracesOperatorMappingV3 = map[v3.FilterOperator]string{
+	v3.FilterOperatorIn:              "IN",
+	v3.FilterOperatorNotIn:           "NOT IN",
+	v3.FilterOperatorEqual:           "=",
+	v3.FilterOperatorNotEqual:        "!=",
+	v3.FilterOperatorLessThan:        "<",
+	v3.FilterOperatorLessThanOrEq:    "<=",
+	v3.FilterOperatorGreaterThan:     ">",
+	v3.FilterOperatorGreaterThanOrEq: ">=",
+	v3.FilterOperatorLike:            "ILIKE",
+	v3.FilterOperatorNotLike:         "NOT ILIKE",
+	v3.FilterOperatorContains:        "ILIKE",
+	v3.FilterOperatorNotContains:     "NOT ILIKE",
+	v3.FilterOperatorExists:          "has(%s%s, '%s')",
+	v3.FilterOperatorNotExists:       "NOT has(%s%s, '%s')",
+}
+
+func getColumnName(key v3.AttributeKey, keys map[string]v3.AttributeKey) (string, error) {
+	key, err := enrichKeyWithMetadata(key, keys)
+	if err != nil {
+		return "", err
+	}
+	if key.IsColumn {
+		return key.Key, nil
+	}
+	filterType, filterDataType := getClickhouseTracesColumnDataTypeAndType(key)
+	return fmt.Sprintf("%s%s['%s']", filterDataType, filterType, key.Key), nil
+}
+
+func getClickhouseTracesColumnDataTypeAndType(key v3.AttributeKey) (v3.AttributeKeyType, string) {
+	filterType := key.Type
+	filterDataType := "string"
+	if key.DataType == v3.AttributeKeyDataTypeFloat64 || key.DataType == v3.AttributeKeyDataTypeInt64 {
+		filterDataType = "number"
+	} else if key.DataType == v3.AttributeKeyDataTypeBool {
+		filterDataType = "bool"
+	}
+	if filterType == v3.AttributeKeyTypeTag {
+		filterType = "TagMap"
+	} else {
+		filterType = "resourceTagsMap"
+		filterDataType = ""
+	}
+	return filterType, filterDataType
+}
+
+func enrichKeyWithMetadata(key v3.AttributeKey, keys map[string]v3.AttributeKey) (v3.AttributeKey, error) {
+	if key.Type == "" || key.DataType == "" {
+		// check if the key is present in the keys map
+		if existingKey, ok := keys[key.Key]; ok {
+			key.IsColumn = existingKey.IsColumn
+			key.Type = existingKey.Type
+			key.DataType = existingKey.DataType
+		} else {
+			return key, fmt.Errorf("key not found to enrich metadata")
```

Comment:
Since we are not deleting keys from the span attributes keys table, if a key is not found, that key will not exist.

Skipping the key would result in unexpected behavior for the user. Eg: He has applied the wrong filter and we are returning data without applying that filter.